File: analyzers/tests/SonarAnalyzer.Rules.Tests/Cases/RedundantNullCheck.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text

Namespace Tests.TestCases
    Class RedundantNullCheck
        Private a As Object
        Private b As Object

        Public Function TestRedundantNullCheck() As Object
            If (a IsNot Nothing And TypeOf a Is RedundantNullCheck) Then ' Noncompliant {{Remove this unnecessary null check; 'TypeOf ... Is' returns false for nulls.}}
'               ^^^^^^^^^^^^^^^
            End If

            If (Not a Is Nothing And TypeOf a Is RedundantNullCheck) Then ' Noncompliant
            End If

            If (a IsNot Nothing AndAlso TypeOf a Is RedundantNullCheck) Then ' Noncompliant
            End If

            If (TypeOf a Is RedundantNullCheck And a IsNot Nothing) Then ' Noncompliant
'                                                  ^^^^^^^^^^^^^^^
            End If

            If (TypeOf a Is RedundantNullCheck And a IsNot Nothing And b IsNot Nothing) Then ' Noncompliant
'                                                  ^^^^^^^^^^^^^^^
            End If

            If (((a) IsNot Nothing) And ((TypeOf (a) Is RedundantNullCheck))) Then ' Noncompliant
            End If

            If (a IsNot Nothing Or TypeOf a Is RedundantNullCheck) Then ' Compliant - not AND operator
            End If

            If (a IsNot Nothing And TypeOf a IsNot RedundantNullCheck) Then ' Compliant
            End If

            If (a IsNot Nothing And Not TypeOf a Is RedundantNullCheck) Then ' Compliant
            End If

            If (a IsNot Nothing And TypeOf b Is RedundantNullCheck) Then ' Compliant
            End If

            If (b IsNot Nothing And TypeOf a Is RedundantNullCheck) Then ' Compliant
            End If

            If (a IsNot Nothing And a IsNot Nothing) Then ' Compliant - not related to this rule
            End If
        End Function

        Public Function TestRedundantInvertedNullCheck() As Object
            If (a Is Nothing Or Not TypeOf a Is RedundantNullCheck) Then ' Noncompliant {{Remove this unnecessary null check; 'TypeOf ... Is' returns false for nulls.}}
'               ^^^^^^^^^^^^
            End If

            If (a Is Nothing OrElse Not TypeOf a Is RedundantNullCheck) Then ' Noncompliant
            End If

            If (a Is Nothing Or TypeOf a IsNot RedundantNullCheck) Then ' Noncompliant
            End If

            If (TypeOf a IsNot RedundantNullCheck Or a Is Nothing) Then ' Noncompliant
'                                                    ^^^^^^^^^^^^

            End If

            If ((TypeOf (a) IsNot RedundantNullCheck) Or ((a) Is Nothing)) Then ' Noncompliant

            End If

            If (a Is Nothing Or Not TypeOf b Is RedundantNullCheck) Then ' Compliant
            End If

            If (b Is Nothing Or Not TypeOf a Is RedundantNullCheck) Then ' Compliant
            End If

            If (a Is Nothing Or a Is Nothing) Then ' Compliant - not related to this rule 
            End If
        End Function
    End Class
End Namespace

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantNullCheck.cs
using System;

namespace Tests.Diagnostics
{
    public class RedundantNullCheck
    {
        public void TestRedundantNullCheck(Object a, RedundantNullCheck b)
        {

            if (a != null && a is RedundantNullCheck) // Noncompliant {{Remove this unnecessary null check; 'is' returns false for nulls.}}
//              ^^^^^^^^^
            {

            }

            if (null != a && a is RedundantNullCheck) // Noncompliant
//              ^^^^^^^^^
            {

            }

            if (a is RedundantNullCheck && a != null) // Noncompliant
//                                         ^^^^^^^^^
            {

            }

            if (a is RedundantNullCheck && a != null && b != null) // Noncompliant
//                                         ^^^^^^^^^
            {

            }

            if (a != null && a is RedundantNullCheck aTyped0) // Noncompliant
            {

            }

            if (a is RedundantNullCheck aTyped1 && aTyped1 != null)
            {

            }

            if (a is RedundantNullCheck aTyped2 && a != null) // Noncompliant
            {

            }

            if (((a) != ((null))) && ((a) is RedundantNullCheck aTyped3)) // Noncompliant
            {

            }

            if (a != null && b != null && b is RedundantNullCheck) // FN
            {

            }

            if (a != null || a is RedundantNullCheck) // Compliant - not AND operator
            {

            }

            if (a != null && !(a is RedundantNullCheck)) // Compliant
            {

            }

            if (a == null && a is RedundantNullCheck) // Compliant - rule ConditionEvaluatesToConstant will raise issue here
            {

            }

            if (a != null && b is RedundantNullCheck) // Compliant
            {

            }

            if (b != null && a is RedundantNullCheck) // Compliant
            {

            }

            if (a != null && a != null) // Compliant - not related to this rule
            {

            }
        }

        public void TestRedundantInvertedNullCheck(Object a, RedundantNullCheck b)
        {

            if (a == null || !(a is RedundantNullCheck)) // Noncompliant {{Remove this unnecessary null check; 'is' returns false for nulls.}}
//              ^^^^^^^^^
            {

            }

            if (!(a is RedundantNullCheck) || a == null) // Noncompliant
            {

            }

            if (a == null || !(a is RedundantNullCheck aTyped1)) // Noncompliant
            {

            }

            if (!(a is RedundantNullCheck aTyped2) || a == null) // Noncompliant
//                                                    ^^^^^^^^^
            {

            }

            if (((!((a) is RedundantNullCheck))) || ((a) == (null))) // Noncompliant
            {

            }

            if (a == null || !(b is RedundantNullCheck)) // Compliant
            {

            }

            if (b == null || !(a is RedundantNullCheck)) // Compliant
            {

            }

            if (a == null || a != null) // Compliant - not related to this rule
            {

            }
        }
    }
}
